fifeforspark/base_modelers.py

```diff
@@ -0,0 +1,430 @@
+from pyspark.sql.functions import lit, when, monotonically_increasing_id
+import findspark
+import pyspark
+from pyspark.mllib.evaluation import BinaryClassificationMetrics
+from pyspark.sql.types import *
+import pandas as pd
+import numpy as np
+from lifelines.utils import concordance_index
+from abc import ABC, abstractmethod
+from typing import Union, Any
+import databricks.koalas as ks
+from collections import OrderedDict
+
+findspark.init()
+
+spark = SparkSession.builder.getOrCreate()
+
+
+def default_subset_to_all(
+    subset: Union[None, pd.core.series.Series], data: pyspark.sql.DataFrame
+) -> pyspark.sql.DataFrame:
+    """Map an unspecified subset to an entirely True boolean mask."""
+    if subset is None:
+        data = data.withColumn('True_mask', lit(True))
+        subset = data.select(data['True_mask'])
+        data = data.drop('True mask')
+        return subset
+    return subset
+
+
+def compute_metrics_for_binary_outcomes(
+    actuals: pyspark.sql.DataFrame, predictions: pyspark.sql.DataFrame,
+    threshold_positive: Union[None, str, float] = 0.5, share_positive: Union[None, str, float] = None
+) -> OrderedDict:
+    actuals = actuals.select(actuals['actuals'].cast(DoubleType()))
+    num_true = actuals.agg({'actuals': 'sum'}).first()[0]
+    total = actuals.count()
+    metrics = OrderedDict()
+    if (num_true > 0) & (num_true < total):
+        predictions = predictions.withColumn(
+            "row_id", monotonically_increasing_id())
+        actuals = actuals.withColumn("row_id", monotonically_increasing_id())
+        preds_and_labs = predictions.join(actuals, predictions.row_id == actuals.row_id).select(
+            predictions.predictions, actuals.actuals)
+        scoresAndLabels = preds_and_labs.rdd.map(tuple)
+        evaluator = BinaryClassificationMetrics(scoresAndLabels)
+        metrics['AUROC'] = evaluator.areaUnderROC
+
+    else:
+        metrics["AUROC"] = np.nan
+
+    # Difficult to do a weighted avg in pyspark, leaving out for now
+    mean_predict = predictions.agg({'predictions': 'mean'}).first()[0]
+    mean_actual = actuals.agg({'actuals': 'mean'}).first()[0]
+    metrics["Predicted Share"] = mean_predict
+    metrics["Actual Share"] = mean_actual
+    # Checks if actuals doesn't have a single row, if so then it's not empty
+    if actuals.first() is not None:
+        if share_positive == 'predicted':
+            share_positive = predictions.agg(
+                {'predictions': 'mean'}).first()[0]
+        if share_positive is not None:
+            threshold_positive = predictions.approxQuantile(
+                'predictions', [1-share_positive], relativeError=.1)[0]
+        elif threshold_positive == 'predicted':
+            threshold_positive = mean_predict
+
+        preds_and_labs = preds_and_labs.withColumn('predictions', when(
+            preds_and_labs.predictions >= threshold_positive, 1).otherwise(0))
+        TP = preds_and_labs.select(((preds_and_labs.predictions == 1) & (
```

Comment:
Can we clean this up? Looks like it could be streamlined visually. If not it's nbd